src/ForeignCallbacks.jl

```diff
@@ -8,97 +8,111 @@ struct Node{T}
 end
 Node(data::T) where T = Node{T}(C_NULL, data)
 
+setnext!(node::Ptr{Node{T}}, next::Ptr{Node{T}}) where {T} =
+    unsafe_store!(convert(Ptr{Ptr{Node{T}}}, node), next)
+
 function calloc(::Type{T}) where T 
     ptr = Libc.malloc(sizeof(T))
     ccall(:memset, Ptr{Cvoid}, (Ptr{Cvoid}, Cint, Csize_t), ptr, 0, sizeof(T))
     return convert(Ptr{T}, ptr)
 end
 
-mutable struct LockfreeQueue{T}
-    @atomic head::Ptr{Node{T}}
-    @atomic tail::Ptr{Node{T}}
-    function LockfreeQueue{T}() where T
-        @assert !Base.ismutabletype(T) && Base.isconcretetype(T) && Base.datatype_pointerfree(T)
-        tmp = calloc(Node{T})
-        new{T}(tmp, tmp)
+"""
+    SingleConsumerStack{T}
+
+A variant of Treiber stack that assumes single-consumer for simple (hopefully)
+correct implementation.
+
+Safety notes: `push!` and `unsafe_push!` on `SingleConsumerStack{T}` can be
+called from multiple tasks/threads. `unsafe_push! can be called from foreign
+threads without the Julia runtime.  Only one Julia task is allowed to call
+`popall!` at any point.  This simplifies the implementation and avoids the use
+of "counted pointer" (and hence 128 bit CAS) that would typically be required
+for general Treiber stack with manual memory management.
```

Comment:
I don't claim I have proof... Someone needs to check it with TLA+ or sth :laughing: